d3a_api_client/redis_aggregator.py

```diff
@@ -182,49 +189,52 @@ def execute_batch_commands(self, is_blocking=True):
+            log_msg['grid_tree'] = '{...}'
+        logging.debug(f"A new message was received. Message information: {log_msg}")
         log_market_progression(message)
-        function = lambda: self.on_event_or_response(message)
-        self.executor.submit(execute_function_util, function=function,
+        self.executor.submit(execute_function_util,
+                             function=lambda: self.on_event_or_response(message),
                              function_name="on_event_or_response")
 
     def calculate_grid_fee(self, start_market_or_device_name: str,
                            target_market_or_device_name: str = None,
-                           fee_type: str = "next_market_fee"):
+                           fee_type: str = "current_market_fee"):
         return self.grid_fee_calculation.calculate_grid_fee(start_market_or_device_name,
                                                             target_market_or_device_name, fee_type)
 
+    def get_uuid_from_area_name(self, name):
+        return get_uuid_from_area_name_in_tree_dict(self.area_name_uuid_mapping, name)
+
+    @buffer_grid_tree_info
     def _on_market_cycle(self, message):
-        self.grid_fee_calculation.handle_grid_stats(message)
-        function = lambda: self.on_market_cycle(message)
-        self.executor.submit(execute_function_util, function=function,
+        self.area_name_uuid_mapping = \
+            create_area_name_uuid_mapping_from_tree_info(self.latest_grid_tree_flat)
+        self.grid_fee_calculation.handle_grid_stats(self.latest_grid_tree)
+        self.executor.submit(execute_function_util, function=lambda: self.on_market_cycle(message),
                              function_name="on_market_cycle")
 
+    @buffer_grid_tree_info
     def _on_tick(self, message):
-        function = lambda: self.on_tick(message)
-        self.executor.submit(execute_function_util, function=function,
+        self.executor.submit(execute_function_util, function=lambda: self.on_tick(message),
                              function_name="on_tick")
 
     @staticmethod
     def _log_trade_info(message):
         logging.info(f"<-- {message.get('buyer')} BOUGHT {round(message.get('energy'), 4)} kWh "
                      f"at {round(message.get('price'), 2)} cents -->")
 
+    @buffer_grid_tree_info
     def _on_trade(self, message):
-        if "content" not in message:
-            # Device message
-            self._log_trade_info(message)
-        else:
-            # Aggregator message
-            for individual_trade in message["content"]:
-                self._log_trade_info(individual_trade)
-        function = lambda: self.on_trade(message)
-        self.executor.submit(execute_function_util, function=function,
+        # Aggregator message
+        for individual_trade in message["trade_list"]:
+            self._log_trade_info(individual_trade)
+        self.executor.submit(execute_function_util, function=lambda: self.on_trade(message),
```

Comment:
So we were trying to handle a `Device message` at the aggregator level, which is not possible , right ?

Reply:
Exactly.